### ISSUE TITLE
Extract transforms from Vega, not Vega-Lite

### ISSUE THREAD

**Issue by domoritz** [vega/vega-lite#4397]
I'd like to propose a different approach to transform extraction. Rather than doing it in Vega-Lite, I think it makes more sense to do it in Vega. 

The advantage is that we don't really need anything because you can just clip off transforms in Vega. We could simplify our codebase. Supporting Vega would also be more general. Another advantage is that we could support interactive charts easier.

The disadvantage is that you have to support more transforms than what Vega-Lite currently uses and you might lose some of the semantics. However, I don't know yet of an example where this would be a problem. 

Concretely, I propose to
* Delete the transform extraction code
* Build a module that unifies the dataflows from multiple Vega specifications
* Build a module that translated Vega to pandas/SQL as a proof of concept. This module does not have to support every operator at first.

**Comment by domoritz** [comment]
cc @kanitw @jheer @Saba9 @tmostak

**Comment by kanitw** [comment]
First of all, I would revise the proposal to change the "Delete the transform extraction code" to "If our prototype proves that this approach works well, delete the transform extraction code" and move it to be the last bullet  (I don't see why we should begin by deleting the working code before prototyping with another approach.).

**Comment by kanitw** [comment]
## Overarching Goals

So let's revisit our goal a bit here to provide more context to other people.

The general goal is to make Vega-Lite more scalable, by enabling users to use Vega-Lite with external data processing engine. 

We see two possible workflows for users:

a. Write transformation specifications totally outside of Vega-Lite and write Vega-Lite specs for the aggregated data.
b. Write Vega-Lite specification with transformations included and use the extract transform method to convert the specification into two parts: 
  1. Transformation specification
  2. Vega-Lite specification without transformation specification in 1., but behave the same otherwise.  

I think a. will actually be quite a common workflow (perhaps even more so than b.).
That said, I observe that implementing b. is actually useful for pushing the support for a. too as we have to design the syntax in the Vega-Lite to support expressing rules that we similarly apply when we extract the transforms.  Thus, for this purpose alone, I think further pushing the `extractTransform` method is a worthwhile investment. 

## Multiple Subproblems?

I think it is also worth breaking down what "transform extraction" means. I think there are multiple levels.

1) Extract transform in Vega-Lite `encoding` (`aggregate`, `bin`, `timeUnit`, and `stack`) to the general `transform` block.

I think this part is quite self contained and is much easier to provide general panda / SQL support (compared to Vega layout transform). We have quite good support for this method already (cover 80% - 90% of what we suppose to do). 
While this is primarily for goal b. above, it also helps us push goal b. forward. Thus, I think we should push this further by considering tracking bin / timeUnit transform #2862, which I think may discover some semantic that we may lose from doing the entire thing in Vega. 

2) Extract `transform` block to be performed outside of Vega/Vega-Lite.  For this part, I can see some benefits of doing it at the Vega level.  

I think it's worth prototyping and think deeply through it.  But I suspect that there are benefits that we gain from doing in Vega-Lite. 


## Pros & Cons

Now let's revisit the pros and cons described above: 

### Pros

> The advantage is that we don't really need anything because you can just clip off transforms in Vega. We could simplify our codebase. Supporting Vega would also be more general. 

Probably true for static (non-interactive) charts 

> Another advantage is that we could support interactive charts easier.

I think we need to elaborate more why this would be easier in Vega, especially that all the selection abstraction that we have will become a collection of hairy datasets and signals. 


### Cons

> you have to support more transforms than what Vega-Lite currently uses

This is probably _not_ a big deal initially if we aim to support this for Vega-Lite users.  Then we can focus on transforms that Vega-Lite currently uses first. 

> you might lose some of the semantics. However, I don't know yet of an example where this would be a problem.

This is the deal breaker for this proposal.  I think we need to further think about this. 

Note: This is quite a hairy topic. Again, I think this kind of discussion can benefit from a more structured RFCs approach.

**Comment by saulshanabrook** [comment]
> We see two possible workflows for users:
> 
> a. Write transformation specifications totally outside of Vega-Lite and write Vega-Lite specs for the aggregated data.
b. Write Vega-Lite specification with transformations included and use the extract transform method to convert the specification into two parts:

To me, it seems that using interactive charts with large remote data sets makes the pre-aggregate approach (a) difficult. For example, if the data is aggregated through a [cross highlight](https://vega.github.io/vega-lite/examples/interactive_concat_layer.html) or you would like to be able to zoom into a heatmap to see a higher resolution, but the data is not client side and you have to call some function to retrieve itt.

In these types of cases option b. (extract transform) seems much more appealing than a. (pre aggregate). Why? Because I imagine there could be an API used with b. such that after some selection happens, the new transformations are extracted and then the backend is called to give data given those transformations. Therefore, to implement a backend you don't have to think about interactions at all, you just have to implement the compilation of Vega Lite transformations to whatever the backend language is like. This approach seems rather [reasonable](https://github.com/vega/vega-lite-transforms2sql) to [implement](https://github.com/Quansight/jupyterlab-omnisci/blob/f8a585579b52e8e440babac5c20047453f469390/jupyterlab_omnisci/altair.py#L166-L184).

Trying to imagine a solution based on a. (pre aggregate) is a lot more fuzzy for me. Then, the backend implementation would then have to know a lot more about events in vega that come up from selections and how to understand what data is needed after those events.

But I also don't know Vega Lite or Vega that intimately, so maybe I am misunderstanding these two pathways.

**Comment by domoritz** [comment]
thanks for the comments. Here are a few more details on the approach and why I think it makes sense. First off, the use case I imagine is support for scalable backends for Altair in data science environments like Jupyterlab with for example Ibis (pandas like API with various backends). 

> I think we need to elaborate more why this would be easier in Vega, especially that all the selection abstraction that we have will become a collection of hairy datasets and signals.

This is really the point that I think is important. Vega-Lite is too high-level to compile to queries because a lot of details gets added by the compiler. @saulshanabrook and I looked at a few examples today and we found that the Vega code actually had a really nice level of detail.

For example, for the [interactive average example](https://vega.github.io/editor/#/examples/vega-lite/selection_layer_bar_month), we see that there are two dataflows. The first one is static and computes a grouped by aggregate. The second one computes an aggregate filtered by the brush (which is just a dataset but the semantics of it are not relevant when we translate to queries). The runtime system needs to rerun this second query when the dataset changes (a simple data change listener). I don't see how this would be easy in vega-Lite.

In both dataflows, we see the extent signals and notice that they are only used in the transforms. We can, therefore, run them completely in the backend. The bin signals are used in the scales and therefore need to be exposed by a backend and injected into the Vega spec by a runtime system.

Supporting interactive charts with scalable backends from Vega-Lite is still an open question. Moreover, we have not even figured out how to really do the transform extraction for Vega-Lite; there are still a number of outstanding issues (https://github.com/vega/vega-lite/issues/4419). By going to vega, we can start implementing support for scalable backends today.

> While this is primarily for goal b. above, it also helps us push goal b. forward. Thus, I think we should push this further by considering tracking bin / timeUnit transform #2862, which I think may discover some semantic that we may lose from doing the entire thing in Vega.

Agreed (I think you meant to say that we push goal forward, though). I think the transform extraction is a good test case but I don't think that should drive design decisions for scalable backends. 

>> you might lose some of the semantics. However, I don't know yet of an example where this would be a problem.

> This is the deal breaker for this proposal. I think we need to further think about this.

I've thought about this for a while and can't find an example where we want the Vega-Lite abstraction to compile to queries for a backend. 

@saulshanabrook I think there is a bit of misunderstanding here. a) and b) are not different approaches but different use cases.

**Comment by kanitw** [comment]
@domoritz Can you summarize your arguments (for and against) doing this in Vega rather than Vega-Lite for transparency and for us to have more time to think about it? 


I have a few of counterpoints for the explanations you have here:

> The advantage is that we don't really need anything because you can just clip off transforms in Vega. We could simplify our codebase. Supporting Vega would also be more general. 

The composite mark codebase needs the [`extractTransformsFromEncoding` method](https://github.com/vega/vega-lite/blob/master/src/compositemark/boxplot.ts#L4), so you can't really remove the code from Vega-Lite.

> For example, for the interactive average example, we see that there are two dataflows. The first one is static and computes a grouped by aggregate. The second one computes an aggregate filtered by the brush (which is just a dataset but the semantics of it are not relevant when we translate to queries). The runtime system needs to rerun this second query when the dataset changes (a simple data change listener). I don't see how this would be easy in Vega-Lite.

Once you run `extractTransformsFromEncoding()`, you will see from a Vega-Lite spec that the aggregation happens _after_ selection filter too. Plus, the selection filter is still in a high-level form.  Assuming we will only port the parts before selection filter to other engine, using Vega-Lite means you don't have to read the underlying expression to understand if a certain filter is a selection filter. 

Thus, I'm not convinced that Vega provides better level of details for the transform part yet.  (It might provide more, but too low-level?) 

(In any cases, Vega definitely provides more details for visual encodings though, but that's not what we're discussing here.)

**Comment by domoritz** [comment]
I'm arguing for extracting the dataflow for a scalable backend from Vega because I think the semantic of Vega is closer to what we want for the optimizations we need to do. For example, [take a histogram chart](https://vega.github.io/editor/#/examples/vega-lite/histogram). In Vega, we split the binning into extent and bin. The extent is used by the bin transform to compute the bin configuration (start, size). The bin transform then creates a signal that is used by the scale. It is fairly easy to reason about how we translate this for a backend. The extent becomes a subquery that creates a signal. When the signal resolves, we compute bin configuration and generate a query to compute the bin counts. We also make the bin config available to the spec. The fact that a signal is needed here doesn't make things much more complicated. 

In a Vega spec, we can see exactly what is going to be executed without much ambiguity. A Vega-Lite spec is more ambiguous and the Vega-Lite compiler has a lot of flexibility about how it compiles the encodings and transforms. This is particularly true for selections where the specific execution can be quite subtle. This means that every person who wants to help developing scalable backends has to know and re-implement the selections behavior. Signals on the other hand are easier to grasp as they map to known concepts (reactive values). I also think that by focusing efforts on Vega we get the advantage of supporting more than just Vega-Lite.

> > The advantage is that we don't really need anything because you can just clip off transforms in Vega. We could simplify our codebase. Supporting Vega would also be more general.

> The composite mark codebase needs the extractTransformsFromEncoding method, so you can't really remove the code from Vega-Lite.

I wasn't aware of this until today. We might be able to simplify the code for this case instead of deleting it.

**Comment by kanitw** [comment]
Thanks  ---  This sounds like a much more convincing example than the one above. 


> I wasn't aware of this until today. We might be able to simplify the code for this case instead of deleting it.

I'm highly doubtful if we can. (Look at the code, there isn't much fat to trim -- file a separate issue if I'm wrong.)  In any case, let's focus on the decision for which dataflow for a scalable backend. 

------

From what I understand, here are the trade-offs:

## Pros for Vega

+ We can see exactly what is going to be executed without much ambiguity (such as the bin case that gets expanded into extent + bin)

+ Get the advantage of supporting more than just Vega-Lite.

## Cons for Vega

- Lack high-level semantic such as knowing about time Unit / selection filter

- Incomplete user story. If we do this in Vega, most back-end will like support only subsets of Vega transforms.  Basically, our user story will become "this is a back-end API for Vega, but you can use this API if you only use this subset of transforms.  In contrast, the user story is stronger if the API is for Vega-Lite and we can pretty much support all the language primitives.  I think there is a middle ground where we do the extraction at the Vega level (suppose we prefer it), but have a wrapper API that takes Vega-Lite as input (but output extracted Vega spec).  Then we can still have good coverage for our input type.  

## Question

- I'm wondering if doing it at the Vega-Lite level means that we need to support only a smaller set of Vega expression.  (Do we need to support only the compile time functions that are already included in the `vega-expression` package, instead of also including the runtime functions.) 

Feel free to reply and edit this point to add more points so we have a good summary

**Comment by kanitw** [comment]
It seems like we'll prototype this in Vega and there is no action item specific to VL.  Let's close this.

**Comment by domoritz** [comment]
Can we leave this open until we add some documentation and an API to do the extract from Vega-Lite? The discussion is extremely valuable.

**Comment by kanitw** [comment]
You can always search for this issue / link back to this issue from your new prototype repo. :)

**Comment by domoritz** [comment]
I can't wait to build a prototype ;-)
